src/charm.py

```diff
@@ -1,115 +1,117 @@
 
-    def _on_config_changed(self, event: ops.ConfigChangedEvent):
-        """Handle changed configuration.
+    def _on_start(self, event: ops.EventBase) -> None:
+        """Handle on start event.
 
-        Change this example to suit your needs. If you don't need to handle config, you can remove
-        this method.
+        Args:
+            event: The event fired on upgrade charm.
+        """
+        self._restart(event)
 
-        Learn more about config at https://juju.is/docs/sdk/config
+    def _restart(self, _: ops.EventBase) -> None:
+        """Restart the jenkins agent charm.
 
-        Args:
-            event: event triggering the handler.
+        Raises:
+            RuntimeError: when the service fails to properly start.
         """
-        # Fetch the new config value
-        log_level = self.model.config["log-level"].lower()
-
-        # Do some validation of the configuration option
-        if log_level in VALID_LOG_LEVELS:
-            # The config is good, so update the configuration of the workload
-            container = self.unit.get_container("httpbin")
-            # Verify that we can connect to the Pebble API in the workload container
-            if container.can_connect():
-                # Push an updated layer with the new config
-                container.add_layer("httpbin", self._pebble_layer, combine=True)
-                container.replan()
-
-                logger.debug("Log level for gunicorn changed to '%s'", log_level)
-                self.unit.status = ops.ActiveStatus()
-            else:
-                # We were unable to connect to the Pebble API, so we defer this event
-                event.defer()
-                self.unit.status = ops.WaitingStatus("waiting for Pebble API")
-        else:
-            # In this case, the config option is bad, so block the charm and notify the operator.
-            self.unit.status = ops.BlockedStatus("invalid log level: '{log_level}'")
-
-    @property
-    def _pebble_layer(self):
-        """Return a dictionary representing a Pebble layer."""
-        return {
-            "summary": "httpbin layer",
-            "description": "pebble config layer for httpbin",
-            "services": {
-                "httpbin": {
-                    "override": "replace",
-                    "summary": "httpbin",
-                    "command": "gunicorn -b 0.0.0.0:80 httpbin:app -k gevent",
-                    "startup": "enabled",
-                    "environment": {
-                        "GUNICORN_CMD_ARGS": f"--log-level {self.model.config['log-level']}"
-                    },
-                }
-            },
-        }
-
-
-if __name__ == "__main__":  # pragma: nocover
-    ops.main.main(IsCharmsTemplateCharm)
+        if not self.model.get_relation(AGENT_RELATION):
+            self.model.unit.status = ops.BlockedStatus("Waiting for relation.")
+            return
+
+        if not self.state.agent_relation_credentials:
```

Comment:
I'm thinking about separation of responsibility and minimizing dependencies as per our architecture guidance here, I think in this case if we want to generalize this code we can either have a function defined in the observer and imported in the charm or the inverse. 

In my opinion the function shouldn't be defined in the observer as it's not really in the scope of the module. And on the other hand the observer should be independent from the charm it's working with (meaning it should only see the charm as `CharmBase` rather than `JenkinsAgentCharm`), and in the future if we switch to a charm library provided by the `jenkins-k8s` charm then it would introduce more problem since we'd have to rewrite a lot of logic to support using the charm library.

A third option would be to create another module and put the function there, but to me this seems to require too much effort just to refactor code.

What do you think? should we leave it as it is or should we go with one of the three approach that I mentioned above?